src/watch.rs

```diff
@@ -195,6 +201,60 @@ impl Watch {
     }
 }
 
+/// Lists the dirs and files in a directory, as two vectors.
+/// Given path must be a readable directory.
+fn list_dir(dir: &Path) -> Result<(Vec<PathBuf>, Vec<PathBuf>), std::io::Error> {
+    let mut dirs = vec![];
+    let mut files = vec![];
+    for entry in std::fs::read_dir(dir)? {
+        let entry = entry?;
+        if entry.file_type()?.is_dir() {
+            dirs.push(entry.path())
+        } else {
+            files.push(entry.path())
+        }
+    }
+    Ok((dirs, files))
```

Comment:
Shouldn't something like this do the trick?
```
    Ok(std::fs::read_dir(dir)?.partition(|e| e.file_type()?.is_dir()))
```

Ah no, can't have that `file_type()?` in there ... that sucks.

Reply:
yeah, in Rust a general rule of thumb I’ve found is that side-effecting stuff is better in imperative loops.